### PR TITLE
FEATURE: API to customize server side composer errors handling in the client side

### PR DESCRIPTION
**Pull request by xfalcox**
This will be used by plugins to handle the client side of their custom post validations without having to overwrite the whole composer save action as it was done in other plugins.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/lib/plugin-api.js [hunk 1]
```diff
@@ -1247,6 +1250,27 @@ class PluginApi {
     Composer.reopen({ beforeSave: method });
   }
 
+  /**
+   * Registers a callback function to handle the composer save errors.
+   * This allows you to implement custom logic that will happen before
+   * the raw error is presented to the user in bootbox modal.
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/lib/plugin-api.js:1256]:
```suggestion
   * the raw error is presented to the user.
```

File: app/assets/javascripts/discourse/app/lib/plugin-api.js [hunk 2]
```diff
@@ -1247,6 +1250,27 @@ class PluginApi {
...
+   * The passed function is expected to return true if the error was handled,
+   * false otherwise.
+   *
+   * Example:
+   *
+   * api.addComposerSaveErrorCallback((error) => {
+   *   if (error == "my_error") {
+   *      //handle error
+   *      return true;
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/lib/plugin-api.js:1265]: I'm a little unsure what `return true/false` means here... does one of them skip displaying the error dialog to the user?

File: app/assets/javascripts/discourse/tests/acceptance/composer-test.js [hunk 3]
```diff
@@ -1106,6 +1106,55 @@ acceptance("Composer - Customizations", function (needs) {
...
+
+    await visit("/");
+    await click("#create-topic");
+    await fillIn("#reply-title", "this title triggers an error");
+    await fillIn(".d-editor-input", "this is the *content* of a post");
+    await click("#reply-control button.create");
+    assert.ok(exists(".dialog-body"), "it pops up an error message");
+    await click(".dialog-footer .btn-primary");
+    assert.ok(!exists(".dialog-body"), "it dismisses the error");
+    assert.ok(exists(".d-editor-input"), "the composer input is visible");
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/tests/acceptance/composer-test.js:1154]: Should we check that `PLUGIN_AGC` message is included in the dialog text here?

File: lib/post_creator.rb [hunk 4]
```diff
@@ -161,7 +161,7 @@ def valid?
       return false
     end
 
-    DiscourseEvent.trigger :before_create_post, @post
+    DiscourseEvent.trigger :before_create_post, @post, @opts
```
Comment (pmusaraj) [inline lib/post_creator.rb:164]: Can we add a test for this and for `before_edit_post`? (These two changes could also be moved to a separate PR, if you would like.)

**Review by pmusaraj — APPROVED** [review]

File: app/assets/javascripts/discourse/app/lib/plugin-api.js (hunk 2 above)
Comment (xfalcox) [reply app/assets/javascripts/discourse/app/lib/plugin-api.js:1265]: Yes, if a single plugin consumer returns true we skip the bootbox, aka the default Discourse error handling in this flow.

File: app/assets/javascripts/discourse/tests/acceptance/composer-test.js (hunk 3 above)
Comment (xfalcox) [reply app/assets/javascripts/discourse/tests/acceptance/composer-test.js:1158]: Handled in 7cd47d84c5

File: lib/post_creator.rb (hunk 4 above)
Comment (xfalcox) [reply lib/post_creator.rb:164]: handled in ad960d66b6
